File: x/agentplane/app/frontend/events.ts
/**
 * Folds shared EventEntries into what the session view renders. This is a projection for one
 * screen, not a second vocabulary. It keeps every entry and the cursor each item, input and turn
 * began at, so the raw view can lay the whole session out as one stream in cursor order with the items in their
 * places.
 */
import { ItemKind, TurnStatus, type Event } from "./x/agentplane/protocol/event_pb";
import type { EventEntry } from "./x/agentplane/protocol/event_log_pb";

export interface Item {
  id: string;
  kind: ItemKind;
  toolName: string;
  /** Streamed text or reasoning, or a tool call's streamed arguments. */
  text: string;
  argumentsJson: string;
  output: string;
  completed: boolean;
  succeeded: boolean | null;
  firstCursor: bigint;
}

export interface Turn {
  id: string;
  status: TurnStatus | null;
  error: string;
  itemIds: string[];
  firstCursor: bigint;
}

export interface InputState {
  id: string;
  state: "confirmed" | "failed";
  detail: string;
  /** The exact native harness message, not an app-owned pending command. */
  text: string;
  /** The turn the harness took it into. */
  turnId: string | null;
  firstCursor: bigint;
}

export interface SessionState {
  harness: "running" | "stopped" | "lost" | null;
  turns: Turn[];
  items: Record<string, Item>;
  inputs: InputState[];
  stderr: string[];
  /** Every source entry the session has produced: what the raw view renders, in `timeline` order. */
  entries: EventEntry[];
  /** The wire's decimal string: a uint64 is not safely a JS number. */
  lastCursor: string;
}

export const EMPTY: SessionState = {
  harness: null,
  turns: [],
  items: {},
  inputs: [],
  stderr: [],
  entries: [],
  lastCursor: "0",
};

/** EventEntry's Event is mandatory in the protocol contract but optional in generated TypeScript. */
export function eventOf(entry: EventEntry): Event {
  if (entry.event === undefined) throw new Error(`event entry ${entry.cursor} has no event`);
  return entry.event;
}

/** What a cursor brought into being, if it brought anything: rendered above that entry's frame. */
export type Row = { kind: "turn"; turn: Turn } | { kind: "input"; input: InputState } | { kind: "item"; item: Item };

export interface TimelineStep {
  entry: EventEntry;
  row: Row | null;
}

/**
 * The session as one stream in cursor order: every entry once, at its own position, carrying the
 * row it started. A row's card therefore sits where the log first mentions it, with the frames that
 * went on filling it following at their own positions — an item's card shows the text it has
 * accumulated by now, which is ahead of the frames below it.
 */
export function timeline(state: SessionState): TimelineStep[] {
  // One entry starts at most one row, so a cursor names at most one.
  const rows = new Map<bigint, Row>();
  for (const turn of state.turns) rows.set(turn.firstCursor, { kind: "turn", turn });
  for (const input of state.inputs) rows.set(input.firstCursor, { kind: "input", input });
  for (const item of Object.values(state.items)) rows.set(item.firstCursor, { kind: "item", item });
  return [...state.entries]
    .sort((left, right) => (left.cursor === right.cursor ? 0 : left.cursor < right.cursor ? -1 : 1))
    .map((entry) => ({ entry, row: rows.get(entry.cursor) ?? null }));
}

/** A turn's items, tool calls and reasoning steps run together into collapsible groups; anything
 * else (an assistant's answer) stands alone and ends the run before it. */
export type ItemGroup = { kind: "single"; item: Item } | { kind: "run"; items: Item[] };

const COLLAPSIBLE: ReadonlySet<ItemKind> = new Set([ItemKind.TOOL_CALL, ItemKind.REASONING]);

export function groupItems(items: Item[]): ItemGroup[] {
  const groups: ItemGroup[] = [];
  for (const current of items) {
    const last = groups.at(-1);
    if (!COLLAPSIBLE.has(current.kind)) {
      groups.push({ kind: "single", item: current });
    } else if (last?.kind === "run") {
      last.items.push(current);
    } else {
      groups.push({ kind: "run", items: [current] });
    }
  }
  return groups;
}

function item(state: SessionState, id: string, firstCursor: bigint): Item {
  return (
    state.items[id] ?? {
      id,
      kind: ItemKind.UNSPECIFIED,
      toolName: "",
      text: "",
      argumentsJson: "",
      output: "",
      completed: false,
      succeeded: null,
      firstCursor,
    }
  );
}

function withItem(state: SessionState, next: Item): SessionState {
  return { ...state, items: { ...state.items, [next.id]: next } };
}

function withInput(state: SessionState, id: string, firstCursor: bigint, update: Partial<InputState>): SessionState {
  const existing = state.inputs.find((input) => input.id === id);
  const inputs = existing
    ? state.inputs.map((input) => (input.id === id ? { ...input, ...update } : input))
    : [
        ...state.inputs,
        { id, state: "confirmed" as const, detail: "", text: "", turnId: null, firstCursor, ...update },
      ];
  return { ...state, inputs };
}

export function reduce(previous: SessionState, entry: EventEntry): SessionState {
  const event = eventOf(entry);
  const state: SessionState = {
    ...previous,
    lastCursor: String(entry.cursor),
    entries: [...previous.entries, entry],
  };
  const observation = event.observation;
  switch (observation.case) {
    case "harnessStarted":
      return { ...state, harness: "running" };
    case "harnessExited":
      return { ...state, harness: "stopped" };
    case "harnessLost":
      return { ...state, harness: "lost" };
    case "harnessStderr":
      return { ...state, stderr: [...state.stderr, observation.value.text] };
    case "harnessUserMessageConfirmed":
      return withInput(state, observation.value.harnessMessageId, entry.cursor, {
        state: "confirmed",
        text: observation.value.text,
        turnId: observation.value.turnId,
        detail: `from ${observation.value.originCommandIds.join(", ")}`,
      });
    case "commandFailed":
      return withInput(state, observation.value.commandId, entry.cursor, {
        state: "failed",
        detail: observation.value.reason,
      });
    case "turnStarted":
      return {
        ...state,
        turns: [
          ...state.turns,
          { id: observation.value.turnId, status: null, error: "", itemIds: [], firstCursor: entry.cursor },
        ],
      };
    case "turnCompleted": {
      const { turnId, status, error } = observation.value;
      return { ...state, turns: state.turns.map((turn) => (turn.id === turnId ? { ...turn, status, error } : turn)) };
    }
    case "itemStarted": {
      const { itemId, kind, toolName } = observation.value;
      const started = { ...item(state, itemId, entry.cursor), kind, toolName };
      const turns = state.turns.map((turn, index) =>
        index === state.turns.length - 1 ? { ...turn, itemIds: [...turn.itemIds, started.id] } : turn
      );
      return { ...withItem(state, started), turns };
    }
    case "textDelta": {
      const current = item(state, observation.value.itemId, entry.cursor);
      return withItem(state, { ...current, text: current.text + observation.value.text });
    }
    case "toolArgumentsDelta": {
      const current = item(state, observation.value.itemId, entry.cursor);
      return withItem(state, { ...current, argumentsJson: current.argumentsJson + observation.value.partialJson });
    }
    case "toolArguments": {
      const current = item(state, observation.value.itemId, entry.cursor);
      return withItem(state, { ...current, argumentsJson: observation.value.argumentsJson });
    }
    case "toolOutputDelta": {
      const current = item(state, observation.value.itemId, entry.cursor);
      return withItem(state, { ...current, output: current.output + observation.value.text });
    }
    case "itemCompleted": {
      const { itemId, outcome } = observation.value;
      const current = item(state, itemId, entry.cursor);
      return withItem(state, {
        ...current,
        completed: true,
        text: outcome.case === "text" ? outcome.value : current.text,
        output: outcome.case === "tool" ? outcome.value.output : current.output,
        succeeded: outcome.case === "tool" ? outcome.value.succeeded : current.succeeded,
      });
    }
    // A Native frame, command admission, and any observation this projection does not model: it
    // renders in raw mode at its own cursor like every other entry, and changes nothing else.
    default:
      return state;
  }
}

File: agentplane/runner/testing/events.py
"""Assertions over the events one attachment saw."""

from __future__ import annotations

from collections.abc import Callable, Sequence

from agentplane.protocol import event_log_pb2, event_pb2

# The generated protocol stubs' own stub chain, which the mypy aspect resolves for direct deps only.
# gazelle:include_dep @pypi//protobuf


def kind(entry: event_log_pb2.EventEntry) -> str:
    return entry.event.WhichOneof("observation") or ""


def of_kind(entries: Sequence[event_log_pb2.EventEntry], name: str) -> list[event_log_pb2.EventEntry]:
    return [entry for entry in entries if kind(entry) == name]


def is_kind(name: str) -> Callable[[event_log_pb2.EventEntry], bool]:
    return lambda entry: kind(entry) == name


def turn_completed(entry: event_log_pb2.EventEntry) -> bool:
    return kind(entry) == "turn_completed"


def items(entries: Sequence[event_log_pb2.EventEntry], item_kind: int) -> list[str]:
    """Ids of the items started with `item_kind`, in order."""
    return [
        entry.event.item_started.item_id
        for entry in of_kind(entries, "item_started")
        if entry.event.item_started.kind == item_kind
    ]


def streamed_text(entries: Sequence[event_log_pb2.EventEntry], item_id: str) -> str:
    return "".join(
        entry.event.text_delta.text
        for entry in of_kind(entries, "text_delta")
        if entry.event.text_delta.item_id == item_id
    )


def completed(entries: Sequence[event_log_pb2.EventEntry], item_id: str) -> event_pb2.ItemCompleted:
    (entry,) = [entry for entry in of_kind(entries, "item_completed") if entry.event.item_completed.item_id == item_id]
    return entry.event.item_completed


def tool_arguments(entries: Sequence[event_log_pb2.EventEntry], item_id: str) -> str:
    (entry,) = [entry for entry in of_kind(entries, "tool_arguments") if entry.event.tool_arguments.item_id == item_id]
    return entry.event.tool_arguments.arguments_json


def assert_contiguous(entries: Sequence[event_log_pb2.EventEntry]) -> None:
    """Source-local cursors are dense and increasing: the log has neither gaps nor duplicates."""
    cursors = [entry.cursor for entry in entries]
    assert cursors == list(range(cursors[0], cursors[0] + len(cursors))), cursors


def assert_sourced(entries: Sequence[event_log_pb2.EventEntry]) -> None:
    """Every event derived from harness output names the Native events it came from."""
    native = {
        entry.origin.sequence
        for entry in of_kind(entries, "native")
        if entry.event.native.direction == event_pb2.DIRECTION_FROM_HARNESS
    }
    derived = (
        "item_started",
        "text_delta",
        "tool_arguments_delta",
        "tool_arguments",
        "tool_output_delta",
        "item_completed",
        "turn_completed",
        "harness_user_message_confirmed",
        "model_changed",
    )
    for entry in entries:
        if kind(entry) in derived:
            assert entry.event.source_sequences, entry
            assert set(entry.event.source_sequences) <= native, entry
